File: medium/39. Combination Sum/39. Combination Sum.ts
function combinationSum(candidates: number[], target: number): number[][] {
    const res: number[][] = []
    const n = candidates.length

    const backtrack = (i: number, items: number[], itemsSum: number) => {
        if (i >= n || itemsSum > target)
            return
        if (itemsSum === target) {
            res.push(items)
            return
        }

        // Use duplicate of the current candidate
        const itemsCopy = [...items]
        itemsCopy.push(candidates[i])
        backtrack(i, itemsCopy, itemsSum + candidates[i])

        // Use the next candidate
        backtrack(i + 1, items, itemsSum)
        return
    }

    backtrack(0, [], 0)

    return res
};

export { }

File: medium/39. Combination Sum/39. Combination Sum.go
package main

func combinationSum(candidates []int, target int) [][]int {
	res := [][]int{}
	n := len(candidates)

	var backtrack func(i, currSum int, items []int)
	backtrack = func(i, total int, items []int) {
		if i >= n || total > target {
			return
		}

		if total == target {
			res = append(res, items)
			return
		}

		// Add current num to total sum
		itemsCopy := append(append([]int{}, items...), candidates[i])
		backtrack(i, total+candidates[i], itemsCopy)

		// Or skip this num and go on
		backtrack(i+1, total, items)
		return
	}

	backtrack(0, 0, []int{})

	return res
}
